### PR TITLE
Add an integration test of chef-client with empty ENV.

### PR DESCRIPTION
**Pull request by ksubrama**
This is an integration test to catch the chef-12.2.0 regression that caused #3153 .  This was one of the identified corrective actions.

### CODE REVIEW

**Comment by ksubrama** [comment]
/cc @btm @jdm @smurawski 

Can one of you /cc chef-client-core as well.

**Comment by jdmundrawala** [comment]
awesome. Just a heads up, I don't think this test will run in travis until it is merged.
:+1: 
cc @chef/client-core

File: spec/integration/client/client_spec.rb [hunk 1]
```diff
@@ -60,6 +60,23 @@ def stop_tiny_server
...
+
+      critical_env_vars = ""
+      %w(PATH RUBYOPT BUNDLE_GEMFILE).each do |option|
+        critical_env_vars += "#{option}=#{ENV[option]} "
+      end
+
+      Chef::Log.info "Bare invocation will have the following load-path."
+      Chef::Log.info shell_out("env -i #{critical_env_vars} ruby -e 'puts $:'").stdout
+      result = shell_out("env -i #{critical_env_vars} #{chef_client} -c \"#{path_to('config/client.rb')}\" -o 'x::default'", :cwd => chef_dir)
+      result.error!
```
Comment (thommay) [inline spec/integration/client/client_spec.rb]: I'd really like to see an expectation here. (And through the rest of this file - there are isolated tests that use expectations, but it's not consistent.)

**Comment by jdm** [comment]
Please remember to cc @jdmundrawala and not me :)

**Comment by ksubrama** [comment]
Dang it!  I was hoping autocomplete would save me but apparently it likes Mozilla.  @jdmundrawala .

**Comment by jdmundrawala** [comment]
@jdm this seems to happen a lot. Is it possible to ignore this project. According to the button, unwatch means you still got at-notifications. Ignored says that you will never get notified.

File: spec/integration/client/client_spec.rb (hunk 1 above)
Comment (ksubrama) [inline spec/integration/client/client_spec.rb]: I think most of them use the implicit test of error! because there's really not much else to verify other than "did this command succeed".  If it didn't, you'd want to see the error log printed (which error! does).  If it did, running a second expect on the return code is redundant because error! already does that.

**Comment by ksubrama** [comment]
I made one change to restrict it to !windows.  env -i is probably not going to work there :-P.

File: spec/integration/client/client_spec.rb [hunk 2]
```diff
@@ -56,7 +58,18 @@ def stop_tiny_server
...
+    end
+
+    it "should complete successfully with no other environment variables", :skip => (Chef::Platform.windows?) do
+      file 'config/client.rb', <<EOM
+local_mode true
+cookbook_path "#{path_to('cookbooks')}"
+EOM
+
+      Chef::Log.info "Bare invocation will have the following load-path."
+      Chef::Log.info shell_out!("env -i #{critical_env_vars} ruby -e 'puts $:'").stdout
```
Comment (danielsdeleo) [inline spec/integration/client/client_spec.rb]: Do we still need these log lines?

Comment (ksubrama) [inline spec/integration/client/client_spec.rb]: Not necessarily - but if that test fails (on some CI environment or elsewhere), I would have no idea how to debug it without either jumping on the box and figuring out the environment or looking at that.  I figured I'd just help my future self by leaving the command to dump the lookup path in there.

Comment (danielsdeleo) [inline spec/integration/client/client_spec.rb]: In that case, I'd do:

```ruby
begin
  run_test_code
rescue
  debugging_statements
  raise
end
```

Comment (ksubrama) [inline spec/integration/client/client_spec.rb]: Done.  Thanks!

**Comment by danielsdeleo** [comment]
Awesome, thanks for doing this! :100: :+1: :ship: :icecream:
